onnxruntime/core/optimizer/gelu_fusion.cc

```diff
@@ -0,0 +1,168 @@
+      continue;
+    }
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(add_node, "Add", {7}) ||
+        add_node.GetExecutionProviderType() != node.GetExecutionProviderType() ||
+        add_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    auto mul_node_itr = add_node.OutputNodesBegin();
+    if (mul_node_itr == add_node.OutputNodesEnd()) {
+      continue;
+    }
+
+    const Node& mul_node = *(mul_node_itr);
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(mul_node, "Mul", {7}) ||
+        mul_node.GetExecutionProviderType() != node.GetExecutionProviderType()) {
+      continue;
+    }
+
+    const Node* mul2_node = nullptr;
+    for (auto iter = mul_node.InputNodesBegin(); iter != mul_node.InputNodesEnd(); ++iter) {
+      if ((*iter).OpType().compare("Mul") == 0) {
+        // find the other input node of Mul
+        mul2_node = &(*iter);
+        break;
+      }
+    }
+    if (mul2_node == nullptr) {
+      continue;
+    }
+
+    std::tuple<bool, NodeArg*, NodeArg*> mul2_input_check;
+    IsInputConstant(graph, *mul2_node, mul2_input_check);
+    if (!std::get<0>(mul2_input_check) || std::get<1>(mul2_input_check) == nullptr || std::get<2>(mul2_input_check) == nullptr) {
+      continue;
+    }
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(*mul2_node, "Mul", {7}) ||
+        mul2_node->GetExecutionProviderType() != node.GetExecutionProviderType() ||
+        mul2_node->GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    const std::vector<NodeArg*> gelu_input_defs{std::get<1>(t)};
+    const std::vector<NodeArg*> gelu_output_defs{const_cast<NodeArg*>(mul_node.OutputDefs()[0])};
+    Node& gelu_node = graph.AddNode(graph.GenerateNodeName("Gelu"),
+                                    "Gelu",
+                                    "fused Gelu subgraphs ",
+                                    gelu_input_defs,
+                                    gelu_output_defs, {}, kMSDomain);
+
+    // Assign provider to this new node. Provider should be same as the provider for old node.
+    gelu_node.SetExecutionProviderType(node.GetExecutionProviderType());
+
+    removed_nodes.push_front(node.Index());
+    removed_nodes.push_front(erf_node.Index());
+    removed_nodes.push_front(add_node.Index());
+    removed_nodes.push_front(mul2_node->Index());
+    removed_nodes.push_front(mul_node.Index());
+    removed_initializers.push_front(std::get<2>(t));
+    removed_initializers.push_front(std::get<2>(add_input_check));
+    removed_initializers.push_front(std::get<2>(mul2_input_check));
+  }
+
+  // Have to remove node in reversed order for now to walk around the issue in RemoveNode
+  for (onnxruntime::NodeIndex removed_node : removed_nodes) {
+    graph.RemoveNode(removed_node);
+  }
+
+  for (auto& tensor_arg : removed_initializers) {
+    if (!HasConsumers(graph, tensor_arg)) {
+      graph.RemoveInitializedTensor(tensor_arg->Name());
+    }
+  }
+
```

Comment:
I assume you delay the check on HasConsumers until after node removal. May be good to capture that in a comment.

Reply:
make sense.